_posts/2021-01-08-consumer-tuning.md

```diff
@@ -0,0 +1,292 @@
+
+![Consumer lag](/assets/images/posts/2021-01-08-consumer-lag.png)
+
+If the bottleneck is in the consumer processes and there are fewer consumers than partitions then adding more consumers to the consumer group subscribed to a topic should help. An important point to make here, though, is that consumer configuration tuning in isolation might not be sufficient to meeting your optimization goals.
+
+Consumer lag growing continually is an indication that the consumer group cannot keep up with the rate of message production. If this happens for long enough, it is possible that the topic retention configurations might mean messages are deleted by the broker before they're read by the consumer. In this case, a short term solution is to increase the `retention.bytes` or `retention.ms` of the topic, but this only puts off the inevitable. The longer term solution is to increase consumer throughput (or slow message production). A more common situation is where the workload is spiky, meaning the consumer lag grows and shrinks. This is only a problem to the extent that the consumer needs to be operating in (near) real time.
+
+### Improving throughput by increasing the minimum amount of data fetched in a request
+
+Use the `fetch.max.wait.ms` and `fetch.min.bytes` configuration properties to set thresholds that control the number of requests from your consumer.
+
+* **`fetch.max.wait.ms`**
+Sets a maximum threshold for time-based batching.
+
+* **`fetch.min.bytes`**
+Sets a minimum threshold for size-based batching.
+
+When the client application polls for data, both these properties govern the amount of data fetched by the consumer from the broker. You can adjust the properties higher so that there are fewer requests, and messages are delivered in bigger batches. You might want to do this if the amount of data being produced is low. Reducing the number of requests from the consumer also lowers the overhead on CPU utilization for the consumer and broker.
+
+Throughput is improved by increasing the amount of data fetched in a batch, but with some cost to latency. Consequently, adjusting these properties lower has the effect of lowering end-to-end latency. We'll look a bit more at targeting latency by increasing batch sizes in the next section.
+
+You can use one or both of these properties. If you use both, Kafka will respond to a fetch request when the first of either threshold is reached.
+
+```properties
+# ...
+fetch.max.wait.ms=500
+fetch.min.bytes=16384
+# ...
+```
+
+## Lowering latency by increasing the maximum batch sizes
```

Comment:
This and some of the following headings suddenly use higher heading levels than the previous ones which used `###` ... I do not think it is intentional, or?

Reply:
I've changed them all to _###_